pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericAvroWriter.java

```diff
@@ -52,4 +67,29 @@ public GenericAvroWriter(Schema schema) {
             this.byteArrayOutputStream.reset();
         }
     }
+
+    /**
+     * This is an adapter from Pulsar GenericRecord to Avro classes.
+     */
+    private class GenericRecordAdapter extends SpecificRecordBase {
+        private GenericRecord message;
+
+        void setCurrentMessage(GenericRecord message) {
+            this.message = message;
+        }
+        @Override
+        public Schema getSchema() {
+            return schema;
+        }
+
+        @Override
+        public Object get(int field) {
+            return message.getField(schema.getFields().get(field).name());
+        }
+
+        @Override
+        public void put(int field, Object value) {
+            throw new UnsupportedOperationException();
+        }
+    }
```

Comment:
Now we generate writer is `GenericDatumWriter`, it can't write `SpecificRecordBase` , we should support SpecificDatumWriter. So, I don't think it a good way to do this.

Reply:
Here we are in a writer and a writer is supposed to only read from the object and not to alter the contents.
So it is good here that we prevent any change.
I cannot understand your point.

The integration tests makes an example of a Souce that is pushing GenericRecord instances and then those messages can be consumed by a standard  Pulsar consumer with AUTO_CONSUME

Reply:
It also can change the message in  GenericRecordAdapter, GenericRecordAdapter seems to have no effect

Reply:
here we are adding an adapter from a Pulsar GenericRecord to SpecificRecordBase but we are interested in implementing only read-only methods.

if you prefer I can leave the implementation of this method empty, the method won't be called by the writer.

If you run the integration test you will see that we are passing thru this code.

Current implementation in master assumes that the GenericRecord is always a GenericAvroRecord, but in case of this new feature it can be any implementation of GenericRecord (and in fact in my case it will be a generic data structure generated dynamically, with a dynamic schema)

Reply:
I understand what you mean,  but `GenericDatumWriter` can write `GenericRecord`. Should we make GenericRecordAdapter implement `GenericRecord` not `SpecificRecordBase`?

Reply:
why we make `GenericRecordAdapter` extend `SpecificRecordBase` not `GenericRecord `, seem `private final GenericDatumWriter<org.apache.avro.generic.GenericRecord> writer;` only can write `GenericRecord` not `SpecificRecordBase`.

Reply:
@congbobo184 I have double checked, and we indeed only need to implement GenericRecord.
The usage of SpecificRecordBase was a left-over of my initial implementation.
I have simplified the code by implementing all of the GenericRecord fields.

good catch !